compiler/rustc_passes/src/dead.rs

```diff
@@ -677,30 +686,87 @@ impl<'tcx> DeadVisitor<'tcx> {
+                            | hir::ItemKind::Union(hir::VariantData::Struct(..), ..),
+                        def_id,
+                        ..
+                    }),
+                    Node::Field(..),
+                ) => Some((*def_id, true)),
+                // enum variants
+                (
+                    Node::Item(hir::Item { kind: hir::ItemKind::Enum(..), def_id, .. }),
+                    Node::Variant(..),
+                ) => Some((*def_id, false)),
+                // variant fields
+                (
+                    Node::Variant(hir::Variant { data: hir::VariantData::Struct(..), id, .. }, ..),
+                    Node::Field(..),
+                ) => Some((id.owner, true)),
+                _ => None,
+            }
+        };
+        if let Some(parent_hir_id) = self.tcx.hir().find_parent_node(id)
+            && let Some(parent_node) = self.tcx.hir().find(parent_hir_id)
+            && let Some(node) = self.tcx.hir().find(id)
+            && let Some((did, is_field)) = field_or_variant_local_def_id(parent_node, node)
+            && !name.as_str().starts_with('_')
+        {
+            self.dead_variants.entry(did).or_default().push(DeadVariant{hir_id: id, span, name, level: lint_level, is_field});
+        }
+    }
+
+    fn warn_multiple_dead_codes(
+        &self,
+        dead_codes: &[(hir::HirId, Span, Symbol)],
         participle: &str,
     ) {
-        if !name.as_str().starts_with('_') {
-            self.tcx.struct_span_lint_hir(lint::builtin::DEAD_CODE, id, span, |lint| {
-                let def_id = self.tcx.hir().local_def_id(id);
-                let descr = self.tcx.def_kind(def_id).descr(def_id.to_def_id());
-                let mut err = lint.build(&format!("{} is never {}: `{}`", descr, participle, name));
-                let hir = self.tcx.hir();
-                if let Some(encl_scope) = hir.get_enclosing_scope(id)
+        if let Some((id, _, name)) = dead_codes.first()
+            && !name.as_str().starts_with('_')
+        {
+            self.tcx.struct_span_lint_hir(
+                lint::builtin::DEAD_CODE,
+                *id,
+                MultiSpan::from_spans(
+                    dead_codes.iter().map(|(_, span, _)| *span).collect(),
+                ),
+                |lint| {
+                    let def_id = self.tcx.hir().local_def_id(*id);
+                    let descr = self.tcx.def_kind(def_id).descr(def_id.to_def_id());
+                    let span_len = dead_codes.len();
+                    let mut err = lint.build(&format!(
+                        "{these}{descr}{s} {are} never {participle}{name}",
+                        these = if span_len > 1 { "these " } else { "" },
+                        s = pluralize!(span_len),
+                        are = pluralize!("is", span_len),
+                        name = if span_len > 1 { String::new() } else { format!(": `{name}`") },
+                    ));
```

Comment:
```suggestion
                    let mut err = lint.build(&format!(
                        "{these}{descr}{s} {names}{are} never {participle}",
                        these = if span_len > 6 { "multiple " } else { "" },
                        s = pluralize!(span_len),
                        names = match dead_codes.iter().map(|(_, _, n)| format!("{n}")).collect::<Vec<_>>() {
                            _ if span_len > 6 => String::new(),
                            [name] => format!("{name} "),
                            [] => unreachable!(),
                            [names.., last] => format!("{} and {last} ", names.join(", "),
                        },
                        are = pluralize!("is", span_len),
                    ));
```

You can simplify by removing the len check for over 6 cases.

Reply:
If we include the ident for the type, we could also add that context to the message:

```
warning: fields `a` and `b` in struct `Foo` are never read
  --> $DIR/issue-85255.rs:7:5
   |
LL | struct Foo {
   |        --- fields in this struct
LL |     a: i32,
   |     ^^^^^^
LL |     pub b: i32,
   |     ^^^^^^^^^^
   |
note: the lint level is defined here
  --> $DIR/issue-85255.rs:4:9
   |
LL | #![warn(dead_code)]
   |         ^^^^^^^^^
```

Reply:
I added the label to the ident.span!